src/serac/numerics/functional/tensor.hpp

```diff
@@ -1170,58 +1172,147 @@ SERAC_HOST_DEVICE bool is_symmetric_and_positive_definite(tensor<double, 3, 3> A
 }
 
 /**
- * @brief Solves Ax = b for x using Gaussian elimination with partial pivoting
- * @param[in] A The coefficient matrix A
- * @param[in] b The righthand side vector b
- * @note @a A and @a b are by-value as they are mutated as part of the elimination
+ * @brief Compute LU factorization of a matrix with partial pivoting
+ * @param[in] A The matrix to factorize
+ * @return x tuple of [P, L, U], with
+ * L the lower triangular factor
+ * U the upper triangular factor
+ * P an array indicating the row permutations due to pivoting.
+ * Row i of A was exchanged with row P[i].
  */
 template <typename T, int n>
-SERAC_HOST_DEVICE constexpr tensor<T, n> linear_solve(tensor<T, n, n> A, const tensor<T, n> b)
+SERAC_HOST_DEVICE constexpr auto lu(const tensor<T, n, n>& A)
 {
+  // BT question for Sam Mish: Why not use the standard library abs?
   constexpr auto abs  = [](double x) { return (x < 0) ? -x : x; };
   constexpr auto swap = [](auto& x, auto& y) {
     auto tmp = x;
     x        = y;
     y        = tmp;
   };
 
-  tensor<double, n> x{};
+  auto           U = A;
+  auto           L = DenseIdentity<n>();
+  tensor<int, n> P(make_tensor<n>([](auto i) { return i; }));
 
   for (int i = 0; i < n; i++) {
     // Search for maximum in this column
-    double max_val = abs(A[i][i]);
+    double max_val = abs(U[i][i]);
 
     int max_row = i;
     for (int j = i + 1; j < n; j++) {
-      if (abs(A[j][i]) > max_val) {
-        max_val = abs(A[j][i]);
+      if (abs(U[j][i]) > max_val) {
+        max_val = abs(U[j][i]);
         max_row = j;
       }
     }
 
-    swap(b[max_row], b[i]);
-    swap(A[max_row], A[i]);
+    swap(P[max_row], P[i]);
+    swap(U[max_row], U[i]);
+  }
 
-    // zero entries below in this column
+  for (int i = 0; i < n; i++) {
+    // zero entries below in this column in U
+    // and fill in L entries
     for (int j = i + 1; j < n; j++) {
-      double c = -A[j][i] / A[i][i];
-      A[j] += c * A[i];
-      b[j] += c * b[i];
-      A[j][i] = 0;
+      double c = U[j][i] / U[i][i];
+      L[j][i]  = c;
+      U[j] -= c * U[i];
+      U[j][i] = 0;
+    }
+  }
+
+  return tuple{P, L, U};
+}
+
+/**
+ * @brief Solves Ax = b for x using Gaussian elimination with partial pivoting
+ * @param[in] A The coefficient matrix A
+ * @param[in] b The righthand side vector b
+ * @return x The solution vector
+ */
+template <typename T, int n>
+SERAC_HOST_DEVICE constexpr tensor<T, n> linear_solve(const tensor<T, n, n> A, const tensor<T, n>& b)
```

Comment:
I was thinking that, too. I wanted to talk you about how best to do it. For example, if `A` is a dual type and `b` isn't, I could promote `b` to dual and call the implemented function again. Slightly less efficient, but doesn't repeat code. Or I could handle each permutation manually.

Reply:
I actually don't want the generic `linear_solve(...)` to handle dual numbers. I wrote a specific implementation for dual numbers that avoids accumulating the derivatives through the factorization (similar to your custom derivative for `inv()`). This will be more efficient (at least for large enough matrices). It's also generic, in the sense that it doesn't depend on the underlying algorithm used for the solve. If we later implement an iterative solver option, this custom derivative will surely be more efficient than the straightforward approach. The problem I see with what I have is that it only works when *both* `A` and `b` hold dual numbers. When I try to use it to handle the case of only one geing dual, I run into compilation problems because my function templates are ambiguous. I see the problem, but I don't know how to solve it. Could you take a look?

Reply:
Update: Sam and I merged the dual and non-dual number implementations while still using the analytic derivative approach that avoids using AD naively on the Gaussian elimination.